### PR TITLE
Service Catalog health check tool

### PR DESCRIPTION
**Pull request by jboyd01**
The objective here is to provide a tool that runs continuously monitoring the health of the Service Catalog deployment.  The health checker exposes metrics (ex error rates, execution times, etc) for Prometheus to scrape.  These metrics in turn can be used to generate alerts.

The health check itself is simple & modeled after our existing e2e tests: perform end to end tests that create a Service Instance and Service Binding and verify the objects are created and have expected status.  Tear down follows and metrics for how long each operation took are collected.  The test runs every 2 minutes by default (override with `--healthcheck-interval`).

The tool can be run as a standalone executable or can be deployed via a helm chart.  At present it is hard coded to use the User Provided Broker and its associated services and plans but this will be parameterized.

Healthcheck was added to the `build` and `images` targets.  It is not yet wired up to push images to a repository and isn't deployed unless you deploy via helm or some other method.

Example metrics scrape:
```
# HELP catalog_health_error_count Number of times the health check ended in error, by error.
# TYPE catalog_health_error_count counter
catalog_health_error_count{error="error getting Broker ups-broker: clusterservicebrokers.servicecatalog.k8s.io \"ups-broker\" not found"} 1
catalog_health_error_count{error="error getting Broker ups-broker: the server could not find the requested resource (get clusterservicebrokers.servicecatalog.k8s.io ups-broker)"} 1
catalog_health_error_count{error="timed out waiting for the condition"} 2
# HELP catalog_health_execution_count Number of times the health check has run.
# TYPE catalog_health_execution_count counter
catalog_health_execution_count 7
# HELP catalog_health_successful_duration_milliseconds Bucketed histogram of processing time (s) of successfully handled requests, by operation.
# TYPE catalog_health_successful_duration_milliseconds histogram
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="100"} 0
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="1000"} 0
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="1500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="2000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="2500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="3000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="3500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="4000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="5000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="6000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="10000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="15000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="20000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="25000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="30000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_deleted",le="+Inf"} 3
catalog_health_successful_duration_milliseconds_sum{method="binding_deleted"} 3670
catalog_health_successful_duration_milliseconds_count{method="binding_deleted"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="100"} 0
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="1000"} 1
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="1500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="2000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="2500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="3000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="3500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="4000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="5000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="6000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="10000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="15000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="20000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="25000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="30000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="binding_ready",le="+Inf"} 3
catalog_health_successful_duration_milliseconds_sum{method="binding_ready"} 2970
catalog_health_successful_duration_milliseconds_count{method="binding_ready"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="100"} 0
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="1000"} 0
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="1500"} 2
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="2000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="2500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="3000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="3500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="4000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="5000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="6000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="10000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="15000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="20000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="25000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="30000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="create_instance",le="+Inf"} 3
catalog_health_successful_duration_milliseconds_sum{method="create_instance"} 4391
catalog_health_successful_duration_milliseconds_count{method="create_instance"} 3
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="100"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="1000"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="1500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="2000"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="2500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="3000"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="3500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="4000"} 0
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="5000"} 2
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="6000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="10000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="15000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="20000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="25000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="30000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="healthcheck_completed",le="+Inf"} 3
catalog_health_successful_duration_milliseconds_sum{method="healthcheck_completed"} 13726
catalog_health_successful_duration_milliseconds_count{method="healthcheck_completed"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="100"} 0
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="500"} 0
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="1000"} 2
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="1500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="2000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="2500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="3000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="3500"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="4000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="5000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="6000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="10000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="15000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="20000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="25000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="30000"} 3
catalog_health_successful_duration_milliseconds_bucket{method="instance_deleted",le="+Inf"} 3
catalog_health_successful_duration_milliseconds_sum{method="instance_deleted"} 2412
catalog_health_successful_duration_milliseconds_count{method="instance_deleted"} 3
```

### CODE REVIEW

File: test/healthcheck/framework/healthcheck.go [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+		}
+	},
+}
+
+var (
+	// A Kubernetes and Service Catalog client
+	kubeClientSet           kubernetes.Interface
+	serviceCatalogClientSet clientset.Interface
+
+	// Namespace in which all test resources should reside
```
Comment (pmorie) [inline test/healthcheck/framework/healthcheck.go]: So, this assumes that the ups broker is deployed, right?

I wonder if we ought to make these things CLI params.

**Comment by pmorie** [comment]
Is there a way that we can easily run this as a make target?

File: test/healthcheck/framework/healthcheck.go (hunk 1 above)
Comment (jboyd01) [reply test/healthcheck/framework/healthcheck.go]: @pmorie, yes, as listed in the description, `At present it is hard coded to use the User Provided Broker and its associated services and plans but this will be parameterized.`

**Comment by pmorie** [comment]
One thing I would like to do in a follow-up is build a 100% no-op broker that we can use this with.

Also one thing to note, I think there should be at least a flavor of this that runs using a namespaced broker, once ns-scoped brokers are done. I would prefer not to have to have a health check broker in the cluster-scoped catalog :)

File: charts/healthcheck/templates/healthcheck-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    metadata:
+      labels:
+        app: {{ template "fullname" . }}
+        chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+        release: "{{ .Release.Name }}"
+        heritage: "{{ .Release.Service }}"
+    spec:
+      serviceAccountName: "service-catalog-healthcheck"
+      containers:
+      - name: healthcheck
```
Comment (pmorie) [inline charts/healthcheck/templates/healthcheck-deployment.yaml]: So, this will serve with the self-signed cert currently, right?

Comment (jboyd01) [reply charts/healthcheck/templates/healthcheck-deployment.yaml]: Right, if existing certificate & key are found (by default it looks in /var/run/service-catalog-healthcheck but this can be overridden with --cert-dir) they will be used, otherwise it will create self signed cert.

File: test/healthcheck/framework/healthcheck.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+	}
+	operationStartTime := time.Now()
+	instance, err = serviceCatalogClientSet.ServicecatalogV1beta1().ServiceInstances(namespace.Name).Create(instance)
+	if err != nil {
+		e = fmt.Errorf("error creating instance: %v", err)
+		glog.Error(e)
+		return e
+	}
+
+	if instance == nil {
```
Comment (luksa) [inline test/healthcheck/framework/healthcheck.go]: When does this happen?

File: test/healthcheck/framework/healthcheck.go [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+	sc, err := serviceCatalogClientSet.ServicecatalogV1beta1().ServiceInstances(namespace.Name).Get(instanceName, metav1.GetOptions{})
+	if err != nil {
+		e = fmt.Errorf("error getting instance: %v", err)
+		return e
+	}
+
+	if sc.Spec.ClusterServiceClassRef == nil {
+		err = errors.New("ClusterServiceClassRef should not be null")
+		glog.Error(err)
+		return err
```
Comment (luksa) [inline test/healthcheck/framework/healthcheck.go]: Personally, I'd introduce a function that creates, logs and returns the error. That would reduce these three lines to a single one. If used throughout the file, it would not only shave off a few dozen lines of code, but also make the rest of the code stand out more.

Maybe even introduce something like `assertNonNil()` to really clean up things.

File: test/healthcheck/framework/metrics.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var registerMetrics sync.Once
+
+const (
+	healthcheckNamespace = "catalog_health" // Prometheus namespace (nothing to do with k8s namespace)
```
Comment (luksa) [inline test/healthcheck/framework/metrics.go]: Maybe call this `prometheusNamespace` then? Would help if someone encounters the variable name in the code below (and thus doesn't see this comment)

File: test/healthcheck/framework/healthcheck.go [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+
+		// Start the HTTP server that enables us to serve /healtz and /metrics.   The  metrics can be pulled,
+		// analyzed and alerted on.
+		err = ServeHttp(options)
+		if err != nil {
+			glog.Errorf("Error starting HTTP: %v", err)
+			os.Exit(1)
+		}
+
+		glog.Infof("Sheduled health checks will be run every %v", options.HealthCheckInterval)
```
Comment (luksa) [inline test/healthcheck/framework/healthcheck.go]: Missing 'c' in Scheduled :)

File: charts/healthcheck/templates/healthcheck-deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
+kind: Deployment
+apiVersion: extensions/v1beta1
```
[the remark this answers is not in the document]
Comment (jboyd01) [reply charts/healthcheck/templates/healthcheck-deployment.yaml]: yes, will do.

File: test/healthcheck/framework/healthcheck.go (hunk 3 above)
Comment (jboyd01) [reply test/healthcheck/framework/healthcheck.go]: Should never happen.

**Comment by jboyd01** [comment]
@luksa thanks for the review, I've addressed all items you raised.

File: test/healthcheck/framework/utils.go [hunk 8]
```diff
@@ -116,3 +117,26 @@ func endpointAvailable(c kubernetes.Interface, namespace, name string) wait.Cond
...
+	}
+	partialFileName := file[context:]
+	format := fmt.Sprintf("...%s:%d: %v", partialFileName, line, msg)
+	e := fmt.Errorf(format, param)
+	glog.Error(e)
+	return e
+}
+
+func logError(msg string) error {
+	return logErrorf(msg, "")
```
Comment (luksa) [inline test/healthcheck/framework/utils.go]: Won't `logErrorf()` now log this line instead of the line that calls `logError()`?

Comment (jboyd01) [reply test/healthcheck/framework/utils.go]: gah, yep, thanks

**Comment by LiliC** [comment]
I am not sure this is a good idea, IMO this is actually harmful because this is artificial load on the service-catalog, masking real usage and potentially hiding actual failure as alerts should always be based on ratio. If there is no metrics from an actual broker, then producing them here is the wrong approach. Individual brokers and service-catalog itself need to have these kind of metrics based on real data alerting should be done.

> HealthCheck is a tool that can be used to monitor the basic health of the Service Catalog deployment. 

I find this useful for end-to-end testing as part of the service-catalog CI, but not something that should be part of service-catalog deployment for an end user. Maybe I misunderstood? :)

**Comment by jboyd01** [comment]
@LiliC thanks for the review.  I agree that if you poll too frequently you will induce artificial load (& once every 2 minutes may be too frequent).   This is meant to be one aspect to a monitoring solution, in addition to metrics exposed directly from the Catalog controller and brokers.

**Comment by cdoan1** [comment]
Hi @jboyd01 , as I understand, this is a standalone tool that exercises service catalog function. It can do its work periodically, and it can collect a set of metrics that can be available to prometheus. Seems like a very useful tool / convention to verify a perf profile of each service catalog driver.

This is basically, what I setup to validate new drivers today. This would save me that time and setup.

👍

File: test/healthcheck/framework/healthcheck.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+	pflag "github.com/spf13/pflag"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var options *HealthCheckServer
+
+func Execute() error {
```
Comment (pmorie) [inline test/healthcheck/framework/healthcheck.go]: We discussed this in the f2f and the major ask here was to make the binary live under cmd.

File: Makefile [hunk 10]
```diff
@@ -209,9 +217,11 @@ verify: .init verify-generated verify-client-gen verify-docs verify-vendor
 	    | grep -v ^pkg/client/ \
 	    | grep -v v1beta1/defaults.go); \
 	  do \
+	   echo Running golint --set_exit_status $$i; \
```
Comment (carolynvs) [inline Makefile]: nit: This should be prefixed with `@` perhaps?

File: charts/healthcheck/README.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+```bash
+$ helm install charts/healthcheck --name healthcheck --namespace healthcheck --set imagePullPolicy=Never --set image=healthcheck:canary
+```
+
+## Uninstalling the Chart
+
+To uninstall/delete the `healthcheck` deployment:
+
+```bash
+$ helm delete healthcheck
```
Comment (carolynvs) [inline charts/healthcheck/README.md]: You will need a `--purge` flag so that the release is deleted. Otherwise if they try to install it again, it will fail because the name is taken.

File: charts/healthcheck/README.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+```
+
+The command removes all the Kubernetes components associated with the chart and
+deletes the release.
+
+## Configuration
+
+The following tables lists the configurable parameters of the HealthCheck
+
+TBD
```
Comment (carolynvs) [inline charts/healthcheck/README.md]: Can you fill in this TBD in this PR?

File: charts/healthcheck/templates/serviceaccounts.yaml [hunk 13]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: List
+items:
+  # The SA for the apiserver
```
Comment (carolynvs) [inline charts/healthcheck/templates/serviceaccounts.yaml]: Comment is a copy/paste error, looks like.

What about removing the list and only having the healthcheck SA defined in this file?

File: charts/healthcheck/values.yaml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+# Default values for Health Check
+# Image to use
+image: quay.io/kubernetes-service-catalog/healthcheck:v0.1.9
+# ImagePullPolicy; valid values are "IfNotPresent", "Never", and "Always"
+imagePullPolicy: Never
```
Comment (carolynvs) [inline charts/healthcheck/values.yaml]: Why is the default Never and not IfNotPresent?

File: cmd/healthcheck/framework/healthcheck.go [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+)
+
+var options *HealthCheckServer
+
+// Execute starts the HTTP Server and runs the health check tasks on a periodic basis
+func Execute() error {
+	pflag.CommandLine.AddGoFlagSet(goflag.CommandLine)
+	options = NewHealthCheckServer()
+	options.AddFlags(pflag.CommandLine)
+	pflag.CommandLine.Set("alsologtostderr", "true")
```
Comment (carolynvs) [inline cmd/healthcheck/framework/healthcheck.go]: 👍 silly glog...

File: cmd/healthcheck/framework/healthcheck.go [hunk 16]
```diff
@@ -0,0 +1,445 @@
...
+	h.createBinding()
+	h.deprovision()
+	h.deleteNamespace()
+
+	if h.frameworkError == nil {
+		ReportOperationCompleted("healthcheck_completed", hcStartTime)
+		glog.V(2).Infof("Successfully ran health check in %v", time.Since(hcStartTime))
+		glog.V(4).Info("") // for readabilty/separation of test runs
+	} else {
+		h.cleanup()
```
Comment (carolynvs) [inline cmd/healthcheck/framework/healthcheck.go]: Would it make sense to have this in a defer so that if for any reason things go wrong, we don't accidentally leave resources around?

File: cmd/healthcheck/framework/healthcheck.go [hunk 17]
```diff
@@ -0,0 +1,445 @@
...
+
+	if sc.Spec.ClusterServiceClassRef == nil {
+		return h.setError("ClusterServiceClassRef should not be null", "")
+	}
+	if sc.Spec.ClusterServicePlanRef == nil {
+		return h.setError("ClusterServicePlanRef should not be null", "")
+	}
+
+	if strings.Compare(sc.Spec.ClusterServiceClassRef.Name, h.serviceclassID) != 0 {
+		return h.setError("ClusterServiceClassRef.Name should not be null", "")
```
Comment (carolynvs) [inline cmd/healthcheck/framework/healthcheck.go]: The error message here, and the next check, are copy/paste errors. They should output the expected value instead.

File: cmd/healthcheck/framework/healthcheck.go [hunk 18]
```diff
@@ -0,0 +1,445 @@
...
+			Status: v1beta1.ConditionTrue,
+		},
+	)
+	if err != nil {
+		return h.setError("binding not ready: %v", err.Error())
+	}
+	ReportOperationCompleted("binding_ready", operationStartTime)
+
+	glog.V(4).Info("Validating that a secret was created after binding")
+	_, err = h.kubeClientSet.CoreV1().Secrets(h.namespace.Name).Get("my-secret", metav1.GetOptions{})
```
Comment (carolynvs) [inline cmd/healthcheck/framework/healthcheck.go]: It would be useful as a follow-on to check that the secret has the expected keys.

File: contrib/examples/prometheus/prometheus.yml [hunk 19]
```diff
@@ -63,7 +63,7 @@ spec:
         - "--storage.tsdb.retention=24h"
         ports:
         - containerPort: 9090
-          protocol: TCP
+          hostPort: 9090
```
Comment (carolynvs) [inline contrib/examples/prometheus/prometheus.yml]: This is just for my own curiosity, but why change from protocol to hostport? It's not immediately obvious to me what that accomplishes.

File: pkg/controller/controller_instance.go [hunk 20]
```diff
@@ -1545,12 +1545,19 @@ func (c *controller) prepareServiceInstanceLastOperationRequest(instance *v1beta
 		return nil, err
 	}
 
+	// allow for nil ServicePlan which happens if user sets to non-existing plan
```
Comment (carolynvs) [inline pkg/controller/controller_instance.go]: I wasn't expecting changes to this file? Was this intentional? If so, I think it should be in a separate PR with more info on what it's fixing.

Comment (jboyd01) [reply pkg/controller/controller_instance.go]: nope, that definitely is a bleed over from another PR, good catch, thanks much @carolynvs !

File: contrib/examples/prometheus/prometheus.yml (hunk 19 above)
Comment (jboyd01) [reply contrib/examples/prometheus/prometheus.yml]: sloppy on my part - temporary for debug/easy of use, I added (well actually replaced protocol which defaults to TCP anyway) hostPort which causes the container to expose a port on your host so you can access the pod's endpoint from the native side (outside of the cluster).  I've reverted this change, thanks for the keen set of eyes!

File: cmd/healthcheck/framework/healthcheck.go [hunk 21]
```diff
@@ -65,7 +65,7 @@ var rootCmd = &cobra.Command{
 			os.Exit(1)
 		}
 
-		// Start the HTTP server that enables us to serve /healtz and /metrics.   The  metrics can be pulled,
+		// Start the HTTP server that enables us to serve /healthz and /metrics.   The  metrics can be pulled,
```
Comment (carolynvs) [inline cmd/healthcheck/framework/healthcheck.go]: 👍 Nice catch!
